### PR TITLE
Auto_ts Handler

### PR DESCRIPTION
**Pull request by pranavvp16**
@martyna-mindsdb here is the PR and the issue we discussed on slack
Screenshot 2023-10-09 at 8 11 00 PM
The above query gives error `Table not found for identifier: t.Time_Period` while trying to predict with the handler
#3499

### CODE REVIEW

**Comment by martyna-mindsdb** [comment]
@pranavvp16 
Can you provide a csv file with sample data you are using please?

**Comment by martyna-mindsdb** [comment]
@pranavvp16 
In the SELECT statement, replace the FROM clause to be like `FROM files.test_sales3 AS t`.

In my local env, it results in this error:

image

File: mindsdb/integrations/handlers/auto_ts_handler/setup.py
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/setup.py:1]: Contents in this file seem duplicated from `__init__.py`. Probably not the culprit, but would be good to fix this and follow the pattern that other ML handlers show.

File: mindsdb/integrations/handlers/auto_ts_handler/requirements.txt
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/requirements.txt:1]: Duplicate? If `dask[distributed]` is enough, then let's remove this.

**Review by paxcema — CHANGES_REQUESTED** [review]
I don't think this integration is ready for review. The fitting process fails quite easily due to how arguments are being handled. Here's an example:

```Start of Fit.....
    Target variable given as = Marketing_Expense
Start of loading of data.....
    Inputs: ts_column = Time_Period, sep = ,, target = ['Marketing_Expense']
    Using given input: pandas dataframe...
    Time_Period column exists in given train data...
    train data shape = (18, 1)
Alert: Could not detect strf_time_format of Time_Period. Provide strf_time format during "setup" for better results.
setting maxlag to a low number = 5
Running Augmented Dickey-Fuller test with paramters:
    maxlag: 5 regression: c autolag: BIC
Data is already stationary
There is no differencing needed in this datasets for VAR model
    No time series plot since verbose = 0. Continuing
Time Interval is given as M
    Correct Time interval given as a valid Pandas date-range frequency...
==================================================
Building Auto SARIMAX Model
==================================================
Running Auto SARIMAX Model...
Exception occurred while building Auto SARIMAX model...
'>' not supported between instances of 'str' and 'int'
    Auto SARIMAX model error: predictions not available.
Best Model is: auto_SARIMAX
    Best Model (Mean CV) Score: inf
--------------------------------------------------
Total time taken: 0 seconds.
--------------------------------------------------
Leaderboard with best model on top of list:
            name  rmse
0  auto_SARIMAX   inf
```

Key part above being `Time Interval is given as M. Correct Time interval given as a valid Pandas date-range frequency...`. That means for this particular argument you need to interpret these strings and build the corresponding date range freq object from Pandas.

@pranavvp16 please use a debugger to verify the training process finishes correctly. Additionally, this will let you understand better where the error at prediction time is happening, exactly.

**Comment by pranavvp16** [comment]
```
Leaderboard with best model on top of list:
            name        rmse
0       Prophet   49.692177
3            ML   99.158912
1  auto_SARIMAX  114.708122
2           VAR  135.979141

```
I have now handled all the inputs correctly and the model train fine on all the types and shows the leaderboard of best model @paxcema

**Comment by pranavvp16** [comment]
Screenshot 2023-10-10 at 8 31 57 PM
just a last fix , Is there anything wrong in the SQL query

**Comment by pranavvp16** [comment]
Screenshot 2023-10-11 at 11 37 56 PM
Screenshot 2023-10-11 at 11 39 40 PM
Screenshot 2023-10-11 at 11 40 10 PM
Screenshot 2023-10-11 at 11 40 19 PM
Screenshots of handler working @paxcema

File: mindsdb/integrations/handlers/auto_ts_handler/README.md
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/README.md:16]: Is this true? Is the user to assume that exogenous variable support is first class, then?

File: mindsdb/integrations/handlers/auto_ts_handler/README.md
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/README.md:14]: `scikit-learn ML` should be renamed to `scikit learn`

File: mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py:62]: This section seems overly verbose. I recommend the following pattern to set default values if the user does not provide custom values:

```python
param = args.get('param_name', default_value)
```

Doing this will let you do without lines 53-60

File: tests/unit/executor_test_base.py
Comment (paxcema) [inline tests/unit/executor_test_base.py:123]: I don't see any unit tests. @ZoranPandovski should unit tests be created before accepting a submission? If so, please add some. Otherwise, please remove these two lines.

**Review by paxcema — CHANGES_REQUESTED** [review]

**Review by paxcema — CHANGES_REQUESTED** [review]
Thanks for the changes @pranavvp16, looking good. While we wait for Zoran's answer regarding tests, there are still a couple unaddressed comments for when you've got time in order for us to move forward.

**Comment by pranavvp16** [comment]
@paxcema Fixed the pending ones directly from [docs](https://github.com/AutoViML/Auto_TS#tips), waiting for @ZoranPandovski

**Comment by ZoranPandovski** [comment]
@pranavvp16, thanks for implementing all of the required changes. Let's add a few tests for create and create_validation. For reference, you can check https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/test_rag.py

**Comment by pranavvp16** [comment]
@ZoranPandovski I have added some unit tests related to create and create_validation

**Comment by ZoranPandovski** [comment]
Please just pull the latest from staging so the OpenAI tests pass

**Comment by pranavvp16** [comment]
@ZoranPandovski done

File: mindsdb/integrations/handlers/auto_ts_handler/README.md
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/README.md:36]: Both `ts_column` and `target` columns should not be passed via `USING`, but through `ORDER BY` and `PREDICT` respectively. Please check this is the case, and also modify the README.

Comment (paxcema) [reply mindsdb/integrations/handlers/auto_ts_handler/README.md:36]: To be clear: examples below show usage of `PREDICT`, but the user should not need to repeat this name in `USING` later on.

File: mindsdb/integrations/handlers/auto_ts_handler/README.md
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/README.md:88]: This should use the `WHERE m.Time_Period > LATEST` syntax. Please refer to other time series handlers to conform to the same behavior.

File: mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py:52]: This dict will contain a `timeseries_settings` key when created using time series syntax.

File: mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py:71]: Get from time series settings map instead of using.

File: tests/unit/ml_handlers/test_auto_ts.py
Comment (paxcema) [inline tests/unit/ml_handlers/test_auto_ts.py:44]: Change to time series specific model creation (i.e. `ORDER BY` clause)

File: tests/unit/ml_handlers/test_auto_ts.py
Comment (paxcema) [inline tests/unit/ml_handlers/test_auto_ts.py:64]: Change to time series specific model creation (i.e. ORDER BY clause)

File: tests/unit/ml_handlers/test_auto_ts.py
Comment (paxcema) [inline tests/unit/ml_handlers/test_auto_ts.py:84]: Change to time series specific model creation (i.e. ORDER BY clause)

File: tests/unit/ml_handlers/test_auto_ts.py
Comment (paxcema) [inline tests/unit/ml_handlers/test_auto_ts.py:104]: Change to time series specific model creation (i.e. ORDER BY clause)

File: tests/unit/ml_handlers/test_auto_ts.py
Comment (paxcema) [inline tests/unit/ml_handlers/test_auto_ts.py:124]: Change to time series specific model creation (i.e. ORDER BY clause)

File: tests/unit/ml_handlers/test_auto_ts.py
Comment (paxcema) [inline tests/unit/ml_handlers/test_auto_ts.py:145]: Change to time series specific model creation (i.e. ORDER BY clause)

File: tests/unit/ml_handlers/test_auto_ts.py
Comment (paxcema) [inline tests/unit/ml_handlers/test_auto_ts.py:42]: Do these tests run correctly on a fresh install? I don't see the `Sales_and_Marketing` file being added in the PR, so I suspect these may fail.

File: tests/unit/ml_handlers/test_auto_ts.py
Comment (paxcema) [inline tests/unit/ml_handlers/test_auto_ts.py:9]: More generally, once you fix the points previously mentioned, please make sure tests actually call the predictor with `JOIN` and `LATEST`. Refer to our docs for examples on how this looks like. That will ensure this is a properly implemented time series integration (as in, it conforms to how others behave, too).

**Review by paxcema — DISMISSED** [review]
Just noticed that there are some pending changes we require before being able to merge this. Thanks for the progress so far @pranavvp16, please take a look and address when you've got time.

**Comment by pranavvp16** [comment]
@paxcema I have added tests which check the predictions of the handler with `JOIN` and also fixed the way input of Time-Series Column was taken using `ORDER BY`. I have also the added the `sales.csv` in the data folder which will support tests

**Review by MinuraPunchihewa — COMMENTED** [review]
Hey @pranavvp16,
I tried creating a model by running your branch locally, but I am getting the following error,
image

Did you come across this?
@dusvyat Any ideas?

**Comment by pranavvp16** [comment]
@MinuraPunchihewa there's no such exception thrown from the create validation method, can you share the console output

**Comment by MinuraPunchihewa** [comment]
@pranavvp16 This is it,
```
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/http/namespaces/sql.py", line 40, in post
    result = mysql_proxy.process_query(query)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/mysql_proxy.py", line 495, in process_query
    executor.query_execute(sql)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/executor.py", line 116, in query_execute
    self.do_execute()
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/executor.py", line 213, in do_execute
    ret = self.command_executor.execute_command(self.query)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/executor_commands.py", line 579, in execute_command
    return self.answer_create_predictor(statement)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/functions.py", line 68, in wrapper
    return func(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/executor_commands.py", line 1551, in answer_create_predictor
    df = self.session.model_controller.create_model(statement, ml_handler)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/interfaces/model/model_controller.py", line 315, in create_model
    predictor_record = ml_handler.learn(**params)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 138, in wrapper
    result = function(*args, **kwargs)
  File "/home/minura/Documents/work/mindsdb/repos/contributor_repos/pranavvp16/mindsdb/./mindsdb/integrations/libs/ml_exec_base.py", line 168, in learn
    self.handler_class.create_validation(
TypeError: create_validation() got an unexpected keyword argument 'handler_storage'
```
It seems to be coming from the core MindsDB code though, however, I am not experiencing this for the other handlers.

**Comment by pranavvp16** [comment]
@MinuraPunchihewa @paxcema can I make a loom video demonstrating the working of handler, cause it works on my device and also works on my ubuntu desktop

**Comment by MinuraPunchihewa** [comment]
@pranavvp16 Yes, please do. That would help.

File: mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py
Comment (ea-rus) [inline mindsdb/integrations/handlers/auto_ts_handler/auto_ts_handler.py:16]: As I look at other handlers code, interface of this function should be:
```python
def create_validation(target, args=None, **kwargs):
```
It started to receive new param handler_storage BTW

**Review by MinuraPunchihewa — COMMENTED** [review]
Hey @pranavvp16,
It seems to work fine now, but there is one little concern that I have, your logic seems to add the suffix _preds to the column that predictions are made on. Is there a reason for this?
I'm just thinking that if I create a model with `PREDICT Sales`, for instance, I would expect to query my predictions using the `Sales` column.

**Comment by pranavvp16** [comment]
fine I'll change that anything else @MinuraPunchihewa

**Comment by MinuraPunchihewa** [comment]
@pranavvp16 That is about it right now.

**Comment by pranavvp16** [comment]
so should I change the prediction column name to sales or keep it as it is ? @MinuraPunchihewa

**Comment by MinuraPunchihewa** [comment]
@pranavvp16 It should be whatever comes after the `PREDICT` keyword.

**Comment by pranavvp16** [comment]
fine I'll make the change

**Comment by dusvyat** [comment]
from my side, thanks for the contribution @pranavvp16 . After you make the suggested change, it is good to go

**Comment by pranavvp16** [comment]
@dusvyat welcome, @MinuraPunchihewa I have changed the prediction column accordingly. So are we ready for merge ?

**Comment by MinuraPunchihewa** [comment]
@pranavvp16 @chandrevdw31 I believe we are ready to merge this.
@pranavvp16 Can you please check and resolve the CI errors that are popping up. I think you have a bunch of dependencies defined in your requirements.txt that are not in use.

By the way, one tiny thing I forgot, can you update the examples in your README with the last change that we did (by removing '_preds')?

**Comment by pranavvp16** [comment]
@MinuraPunchihewa thanks for the suggestion I have changed `README.md` and `requirements.txt` accordingly. @chandrevdw31 let's get this merged

**Comment by chandrevdw31** [comment]
@MinuraPunchihewa @pranavvp16 will merge as soon as checks are done

**Review by chandrevdw31 — APPROVED** [review]

**Comment by pranavvp16** [comment]
@chandrevdw31 @MinuraPunchihewa  I think the requirements check fails due to version conflicts of auto-ts. What to do in such scenario

**Comment by pranavvp16** [comment]
@MinuraPunchihewa Looking at the failing tests what I infer is it tries to check whether the requirements match import but fails as requirements has `auto-ts` and import is `auto_ts`. If this is the issue it needs to be fixed in the CI. Please let me know if it isn't this and how can I fix it

**Comment by pranavvp16** [comment]
After a work around I found the issue is with `auto-ts`, I will make a PR in the repository and fix the issue

**Comment by MinuraPunchihewa** [comment]
Hey @pranavvp16,
If you've made the PR, you can link it here. Just to keep a track of it.

**Comment by pranavvp16** [comment]
till then can we `pip install <my_fork>` which is a fixed version of auto_ts ?. I'll change it as soon it is fixed in the main auto_ts repo

**Comment by MinuraPunchihewa** [comment]
@pranavvp16 I believe we have made the decision to avoid installing forks at the moment, unfortunately.
@StpMax Could you advice us on how best go about this?

**Comment by StpMax** [comment]
@MinuraPunchihewa you mean issue with `auto-ts`->`auto_ts`? here is small PR to fix it: https://github.com/mindsdb/mindsdb/pull/8417

**Comment by MinuraPunchihewa** [comment]
@StpMax Thanks, Max.
@pranavvp16 I see that your other PR has been merged. I guess we now need to wait for the new release.
By the way, there also seems to be a small issue that Flake has picked up. Can you please fix that?

**Comment by pranavvp16** [comment]
@MinuraPunchihewa auto-ts has released the latest version and I was also fixed the flake8 error

**Comment by MinuraPunchihewa** [comment]
Hey @pranavvp16,
That's great! A couple of CI checks are still failing. Can you please have a look?
@paxcema If you have any other comments on this integration, they are certainly welcome!

**Comment by MinuraPunchihewa** [comment]
Hey @pranavvp16,'
I believe the staging branch of your fork is behind staging branch of the original `mindsdb` repository. Can you please sync your fork and then merge staging into your branch?

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]
@pranavvp16 I'm really sorry for requesting you make changes again, but can we please do two things (I believe these will resolve the CI issues as well),
1. Let's remove the `setup.py` file; it is not required. After speaking to @paxcema, I understood that this was part of the old design for integrations and is not in practice currently.
2. Let's discard the changes that you've made to `executor_test_base.py`; after speaking to @ea-rus I understood that the line `r = db.Integration(name="auto_ts", data={}, engine="auto_ts")` is basically used is somewhat the equivalent for the `CREATE ML_ENGINE` command. If you discard your changes, it will fix the CI issue where there is no new line at the end of the file.
3. Let's improve your tests a little bit; there is a BaseMLAPITest class that you can inherit from to make your code a little more DRY. Some of the common functions like `wait_predictor()` and `run_sql()` are implemented here. Further, let's add a `setup_method()` function to your test class, so that you can put some of the common code in there including the `self.run_sql("create database proj")` line **AND** the `CREATE ML_ENGINE` command which we have removed in 2 above. You can refer the [tests for Anthropic](https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/base_ml_test.py) to get a better idea of this.

Again, apologies for asking for more changes after the PR was approved.

**Comment by pranavvp16** [comment]
The test you shared doesn't implement any of the methods instead it has methods like `run_sql` and `wait_predictor`. I was asked by Zoran to write the tests where he shared me this [test](https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/test_rag.py) example and asked to add create and validation tests only, later I also added predict tests too. Please share me a better example that implements the class your talking about @MinuraPunchihewa

**Pull request closed without merging**
